### PR TITLE
Feature/register user

### PR DESCRIPTION
**Pull request by NikoBergemann**
### Reason for this PR
- Lagom needs to be able to register users for matriculated students and for teachers to add them to the system and enable them to make chaincode transactions of their own.

### Changes in this PR
- provide RegistrationManager to enable Registration

### CODE REVIEW

File: src/main/scala/de/upb/cs/uc4/hyperledger/utilities/EnrollmentManager.scala [hunk 1]
```diff
@@ -13,31 +13,31 @@ object EnrollmentManager {
...
-      // enroll my user
-      val enrollmentRequestTLS = new EnrollmentRequest
-      enrollmentRequestTLS.addHost("localhost")
-      enrollmentRequestTLS.setProfile("tls")
+      println(s"[DEBUG] :: Try to get the identity for the user $username.")
+
+      val caClient = CAClientManager.getCAClient(caURL, tlsCert)
+      println("[DEBUG] :: Retrieved CAClient")
+
+      val enrollmentRequestTLS = EnrollmentManager.prepareEnrollmentRequest("localhost", "tls")
```
Comment (matthias-geuchen) [inline src/main/scala/de/upb/cs/uc4/hyperledger/utilities/EnrollmentManager.scala:25]: Not sure about addressing in kubernetes (and if peers check hostnames at all), but if we have different organizations running peers in different clusters, the host should probably be configurable, right?

File: src/main/scala/de/upb/cs/uc4/hyperledger/utilities/RegistrationManager.scala [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+import org.hyperledger.fabric.gateway.{ Identities, X509Identity }
+import org.hyperledger.fabric.sdk.{ Enrollment, User }
+import org.hyperledger.fabric_ca.sdk.{ HFCAClient, RegistrationRequest }
+
+object RegistrationManager {
+
+  @throws[Exception]
+  def register(tlsCert: Path, caURL: String, userName: String, adminName: String, adminWalletPath: Path, affiliation: String = "org1"): String = {
+    // retrieve Admin Identity ad a User
```
Comment (matthias-geuchen) [inline src/main/scala/de/upb/cs/uc4/hyperledger/utilities/RegistrationManager.scala:15]: Typo
```suggestion
    // retrieve Admin Identity as a User
```

**Review by matthias-geuchen — APPROVED** [review]

File: src/main/scala/de/upb/cs/uc4/hyperledger/utilities/EnrollmentManager.scala (hunk 1 above)
Comment (NikoBergemann) [reply src/main/scala/de/upb/cs/uc4/hyperledger/utilities/EnrollmentManager.scala:25]: created an issue

https://github.com/upb-uc4/hlf-api/issues/38

File: src/main/scala/de/upb/cs/uc4/hyperledger/utilities/EnrollmentManager.scala [hunk 3]
```diff
@@ -13,31 +13,31 @@ object EnrollmentManager {
     val wallet = WalletManager.getWallet(walletPath)
 
     // check if user already exists in my wallet
-    if (wallet.list.contains(username)) {
-      println(s"An identity for the user $username already exists in the wallet.")
+    if (WalletManager.containsIdentity(wallet, username)) {
+      println(s"[DEBUG] :: An identity for the user $username already exists in the wallet.")
```
Comment (david-buderus) [inline src/main/scala/de/upb/cs/uc4/hyperledger/utilities/EnrollmentManager.scala:17]: You should think about a Logger.

**Review by david-buderus — APPROVED** [review]

**Review by ldklenner — COMMENTED** [review]
Are these changes documented in the `hlf-api-matriculation.md`?

File: src/main/scala/de/upb/cs/uc4/hyperledger/utilities/EnrollmentManager.scala (hunk 3 above)
Comment (NikoBergemann) [reply src/main/scala/de/upb/cs/uc4/hyperledger/utilities/EnrollmentManager.scala:17]: Issue created

https://github.com/upb-uc4/hlf-api/issues/39

**Comment by NikoBergemann** [comment]
> Are these changes documented in the `hlf-api-matriculation.md`?

We should probably investigate & discuss what parts of the scala documentation should be duplicated in the API - docs
Regarding the specific matriculation-API, nothing has changed. We just provide additional functionality that has nothing to do with matriculation at all.
I think, we should add some Help-Documentation describing use-cases and workflows for interacting with the "Hyperledger-Fabric"-network in general. Maybe extend the "KnowledgeBase" in this regard.

File: src/main/scala/de/upb/cs/uc4/hyperledger/utilities/RegistrationManager.scala [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+import java.util
+
+import org.hyperledger.fabric.gateway.{ Identities, X509Identity }
+import org.hyperledger.fabric.sdk.{ Enrollment, User }
+import org.hyperledger.fabric_ca.sdk.{ HFCAClient, RegistrationRequest }
+
+object RegistrationManager {
+
+  @throws[Exception]
+  def register(tlsCert: Path, caURL: String, userName: String, adminName: String, adminWalletPath: Path, affiliation: String = "org1"): String = {
```
Comment (jonas-thiele) [inline src/main/scala/de/upb/cs/uc4/hyperledger/utilities/RegistrationManager.scala:14]:
```suggestion
  def register(tlsCert: Path, caURL: String, userName: String, adminName: String, adminWalletPath: Path, affiliation: String): String = {

```

File: src/main/scala/de/upb/cs/uc4/hyperledger/utilities/RegistrationManager.scala [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    // 1.2.840.10045.4.3.2
+
+    // println("[DEBUG] ::  TEST: " + identity.getCertificate.getSigAlgParams)
+    // null
+
+    // println("[DEBUG] ::  TEST: " + identity.getCertificate.getSignature)
+    // [B@5d6f35bf
+
+    // identity.getCertificate.getSubjectAlternativeNames.forEach(a => println("[DEBUG] ::  TEST: " + a))
+    // [2, localhost]
```
Comment (jonas-thiele) [inline src/main/scala/de/upb/cs/uc4/hyperledger/utilities/RegistrationManager.scala:83]:
```suggestion
```

Do we still need these?

File: src/main/scala/de/upb/cs/uc4/hyperledger/utilities/RegistrationManager.scala [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
```
Comment (jonas-thiele) [inline src/main/scala/de/upb/cs/uc4/hyperledger/utilities/RegistrationManager.scala:67]:
```suggestion
```

File: src/test/scala/de/upb/cs/uc4/hyperledger/tests/RegistrationTests.scala [hunk 7]
```diff
@@ -0,0 +1,36 @@
+package de.upb.cs.uc4.hyperledger.tests
+
+import de.upb.cs.uc4.hyperledger.testBase.TestBase
+import de.upb.cs.uc4.hyperledger.utilities.{ EnrollmentManager, RegistrationManager, WalletManager }
+
+class RegistrationTests extends TestBase {
+
+  "The registrationManager" when {
+    "beginning registration" should {
+      "work as expected" in {
```
Comment (jonas-thiele) [inline src/test/scala/de/upb/cs/uc4/hyperledger/tests/RegistrationTests.scala:10]:
```suggestion
      "not throw exceptions" in {
```

**Review by jonas-thiele — CHANGES_REQUESTED** [review]
looks good to me

**Review by jonas-thiele — APPROVED** [review]

**Review by ldklenner — APPROVED** [review]

**Pull request merged**